### PR TITLE
Dspeianu/java warm up/development

### PR DESCRIPTION
**Pull request by speianudana**

### CODE REVIEW

File: src/main/java/com/endava/internship/warmup/domain/service/ArrayProcessorWithForLoops.java [hunk 1]
```diff
@@ -0,0 +1,354 @@
...
+public class ArrayProcessorWithForLoops implements ArrayProcessor {
+
+    /**
+     * Return true if there are no numbers that divide by 10
+     *
+     * @param input non-null immutable array of ints
+     */
+    @Override
+    public boolean noneMatch(final int[] input) {
+        int numberOfAppearance = 0;
```
Comment (vcontu) [inline src/main/java/com/endava/internship/warmup/domain/service/ArrayProcessorWithForLoops.java]: not necessary

File: src/main/java/com/endava/internship/warmup/domain/service/ArrayProcessorWithForLoops.java [hunk 2]
```diff
@@ -0,0 +1,354 @@
...
+
+        for (int num : input) {
+            if (num % 10 == 0) {
+                numberOfAppearance++;
```
Comment (vcontu) [inline src/main/java/com/endava/internship/warmup/domain/service/ArrayProcessorWithForLoops.java]: return false

File: src/main/java/com/endava/internship/warmup/domain/service/ArrayProcessorWithForLoops.java [hunk 3]
```diff
@@ -0,0 +1,354 @@
...
+            }
+        }
+
+        return numberOfAppearance == 0;
```
Comment (vcontu) [inline src/main/java/com/endava/internship/warmup/domain/service/ArrayProcessorWithForLoops.java]: return true

File: src/main/java/com/endava/internship/warmup/domain/service/ArrayProcessorWithForLoops.java [hunk 4]
```diff
@@ -0,0 +1,354 @@
...
+                            IntPredicate predicate) {
+        int nrOfMatches = 0;
+        for (String stringValue : input) {
+            int intValue = function.applyAsInt(stringValue);
+            if (predicate.test(intValue)) {
+                nrOfMatches++;
+            }
+        }
+
+        return nrOfMatches == input.length;
```
Comment (vcontu) [inline src/main/java/com/endava/internship/warmup/domain/service/ArrayProcessorWithForLoops.java]: same as above

File: src/main/java/com/endava/internship/warmup/domain/service/ArrayProcessorWithForLoops.java [hunk 5]
```diff
@@ -0,0 +1,354 @@
...
+    @Override
+    public int[] copyValues(int[] input, int startInclusive, int endExclusive) throws IllegalArgumentException {
+        if (endExclusive <= startInclusive || startInclusive < 0) {
+            throw new IllegalArgumentException();
+        }
+        int newArrayLength = endExclusive - startInclusive;
+        int[] newArray = new int[newArrayLength];
+        int index = 0;
+        for (int i = startInclusive; i <= newArrayLength; i++) {
+            if (i >= startInclusive && i < endExclusive) newArray[index++] = input[i];
```
Comment (vcontu) [inline src/main/java/com/endava/internship/warmup/domain/service/ArrayProcessorWithForLoops.java]: I've seen often in a convention at Endava, to use the syntax 
if () { 
} 
for some reason, people consider that no brackets {} version makes it a bit less readable even if you execute only one statement when the if condition is true. Personally i am indifferent, but at real projects they might require of you to always use brackets {}

File: src/main/java/com/endava/internship/warmup/domain/service/ArrayProcessorWithForLoops.java [hunk 6]
```diff
@@ -0,0 +1,354 @@
...
+     * Replace even index values with their doubles and odd indexed elements with their negative
+     *
+     * @param input non-null immutable array of ints
+     * @return new array with changed elements
+     */
+    @Override
+    public int[] replace(final int[] input) {
+        int[] newArray = new int[input.length];
+        for (int i = 0; i < input.length; i++) {
+            if (i % 2 == 0) newArray[i] = input[i] * 2;
```
Comment (vcontu) [inline src/main/java/com/endava/internship/warmup/domain/service/ArrayProcessorWithForLoops.java]: please always place the statement after condition on new line.

File: src/main/java/com/endava/internship/warmup/domain/service/ArrayProcessorWithForLoops.java [hunk 7]
```diff
@@ -0,0 +1,354 @@
...
+        return newArray;
+    }
+
+    /**
+     * Find the second max value in the array
+     *
+     * @param input non-null immutable array of ints
+     */
+    @Override
+    public int findSecondMax(final int[] input) {
```
Comment (vcontu) [inline src/main/java/com/endava/internship/warmup/domain/service/ArrayProcessorWithForLoops.java:114]: I admit the tests where not very comprehensive and did not cover all the possible edge cases.  After reviewing this solution, i tried to run it with this new test:

`    @Test
    void findSecondMax2() {
        final int[] inputArray = {Integer.MIN_VALUE, Integer.MIN_VALUE, Integer.MAX_VALUE, Integer.MAX_VALUE};
        assertThat(testProcessor.findSecondMax(inputArray)).isEqualTo(Integer.MIN_VALUE);
    }
`

the test passed, but i received multiple messages: 

There is no second largest value.
There is no second largest value.
There is no second largest value.
There is no second largest value.

Please review this method

File: src/main/java/com/endava/internship/warmup/domain/service/ArrayProcessorWithForLoops.java [hunk 8]
```diff
@@ -0,0 +1,354 @@
...
+
+    /**
+     * Return in reverse first negative numbers, then positive numbers from array
+     *
+     * @param input non-null immutable array of ints.
+     * @return example: input {3, -5, 4, -7, 2 , 9}
+     * result: {-7, -5, 9, 2, 4, 3}
+     */
+    @Override
+    public int[] rearrange(final int[] input) {
```
Comment (vcontu) [inline src/main/java/com/endava/internship/warmup/domain/service/ArrayProcessorWithForLoops.java:142]: Unfortunately I understood wrongly the AssertJ assetion `assertThat().containsOnly()` from the test for this method. It verifies only the presence of the elements, but not their exact order. The working test should be:
`@Test
    void rearrange() {
        final int[] initialArray = {3, -5, 4, -7, 2, 9};
        final int[] expectedArray = {-7, -5, 9, 2, 4, 3};

        assertThat(testProcessor.rearrange(initialArray)).containsExactly(-7, -5, 9, 2, 4, 3);
    }

with this kind of test it will fail. Please review the example from Java Doc

File: src/main/java/com/endava/internship/warmup/domain/service/ArrayProcessorWithForLoops.java [hunk 9]
```diff
@@ -0,0 +1,354 @@
...
+     *
+     * @param input  first non-null array
+     * @param input2 second non-null array
+     * @return new array containing all elements sorted from input and input2
+     * @throws IllegalArgumentException if either input or input are not sorted ascending
+     */
+    @Override
+    public int[] mergeSortedArrays(int[] input, int[] input2) throws IllegalArgumentException {
+        int resultLength = input.length + input2.length;
+        int[] firstArray = sortArray(input);
```
Comment (vcontu) [inline src/main/java/com/endava/internship/warmup/domain/service/ArrayProcessorWithForLoops.java]: the arrays are already sorted.

**Review by vcontu — CHANGES_REQUESTED** [review]
